### PR TITLE
feat(): support buffer local opt-in/out

### PR DESCRIPTION
**Pull request by mhartington**
An attempt at supporting #27 

@llimllib would you be up for testing this out?

### CODE REVIEW

**Comment by llimllib** [comment]
Sure thing. Am I correct that `:let b:formatter_format_on_save=0` is what you want me to test? Any other tests you'd like me to try?

**Comment by mhartington** [comment]
Yep! Still trying to think of a good variable name..but this will do for now

**Comment by llimllib** [comment]
1. apologies for the slowness!
2. `Formatter: Finished running black
...are/nvim/plugged/formatter.nvim/lua/formatter/format.lua:122: Expected lua table function: builtin#18 ...are/nvim/plugged/formatter.nvim/lua/formatter/format.lua:122: Expected lua table`

**Comment by llimllib** [comment]
I updated neovim, which didn't change anything. It's possible that my `black` config is messed up?

```
local black = function()
  return {exe = "black", stdin = true}
end

require('formatter').setup({
  logging = false,
  filetype = {
    python = { black },
  }
})

vim.api.nvim_exec([[
augroup FormatAutogroup
  autocmd!
  autocmd BufWritePost *.py FormatWrite
augroup END
]], true)
```

**Comment by mhartington** [comment]
Yeah, your black config is wrong, it should be 


```
local black = function()
  return {exe = "black", args = {"-q", "-"}, stdin = true}
end

```

File: lua/formatter/util.lua [hunk 1]
```diff
@@ -89,4 +89,9 @@ function util.fireEvent(event)
   vim.api.nvim_command(cmd)
 end
 
+function util.getBufVar(buf, var)
+  local result = pcall(vim.api.nvim_buf_get_var, buf, var)
+  if not result then return true end
+  return vim.api.nvim_buf_get_var(buf, var)
```
Comment (lukas-reineke) [inline lua/formatter/util.lua:95]: nitpick, but defaulting to `true` when there is no result is a bit confusing. Might be nicer to have the default as argument to `getBufVar`?

Also, `pcall` returns a tuple with status and result. So you don't have to call `nvim_buf_get_var` twice

```lua
local ok, result = pcall(vim.api.nvim_buf_get_var, buf, var)
if ok then
    return result
end
```

**Comment by lukas-reineke** [comment]
I think the`formatter_format_on_save` variable name is a bit weird :thinking: 

All it does is skip any formatting when false right? It's unrelated to saving.

File: lua/formatter/util.lua (hunk 1 above)
Comment (mhartington) [reply lua/formatter/util.lua:95]: pcall will not return the result if the variable is not set. 

So if we went with your suggestions, the status and result when the buffer-local variable is not set will look something like this.

```
status false
result Key not found: formatter_format_on_save
```

Probably simpler to just do 

```
function util.getBufVar(buf, var)
  local status, result = pcall(vim.api.nvim_buf_get_var, buf, var)
  if status then return result end
  return nil
end
```

then when we try to get the variable, we can do 

```
  local buf_skip_format = util.getBufVar(bufnr, 'formatter_format_on_save') or false
```

**Comment by mhartington** [comment]
Yeah, the name is odd. Open to suggestions on it

File: lua/formatter/util.lua (hunk 1 above)
Comment (lukas-reineke) [reply lua/formatter/util.lua:95]: Yes this is what I meant :+1: 

or

```lua
function util.getBufVar(buf, var, default)
  local status, result = pcall(vim.api.nvim_buf_get_var, buf, var)
  if status then return result end
  return default
end
```

but either way is totally fine

**Comment by lukas-reineke** [comment]
`formatter_skip_buf` makes sense to me

**Comment by mhartington** [comment]
Just pushed some updates, will merge and merge in #18 as well

**Pull request merged**